src/services/blocks/BlocksService.ts

```diff
@@ -450,6 +469,153 @@ export class BlocksService extends AbstractService {
+		getFeeByEvent: boolean
+	) {
+		const { api } = this;
+		const { class: dispatchClass, partialFee } =
+			await api.rpc.payment.queryInfo(extrinsicHex, hash);
+
+		/**
+		 * Check if we should retrieve the partial_fee from the Events
+		 */
+		let fee: Balance | string = partialFee;
+		let error: string | undefined;
+		if (getFeeByEvent) {
+			const feeInfo = this.getPartialFeeByEvents(events, partialFee);
+			fee = feeInfo.partialFee;
+			error = feeInfo.error;
+		}
+
+		return {
+			partialFee: fee,
+			dispatchClass,
+			error,
+		};
+	}
+
+	/**
+	 * This searches through an extrinsics given events to see if there is a partialFee
+	 * within the data. If the estimated partialFee is within a given difference of the
+	 * found fee within the data than we return that result.
+	 *
+	 * The order of the events we search through are:
+	 * 1.Balances::Event::Withdraw
+	 * 2.Treasury::Event::Deposit
+	 * 3.Balances::Event::Deposit
+	 *
+	 * @param events The events to search through for a partialFee
+	 * @param partialFee Estimated partialFee given by `payment::queryInfo`
+	 */
+	private getPartialFeeByEvents(
+		events: ISanitizedEvent[],
+		partialFee: Balance
+	): { partialFee: string; error?: string } {
+		// Check Event:Withdraw event for the balances pallet
+		const withdrawEvent = this.findEvent(events, 'balances', Event.withdraw);
+		if (withdrawEvent.length > 0 && withdrawEvent[0].data) {
+			const dataArr = withdrawEvent[0].data.toJSON();
+			if (Array.isArray(dataArr)) {
+				const fee = (dataArr as Array<number>)[dataArr.length - 1];
+
+				// The difference between values is 00.00001% or less so they are alike.
+				if (this.areFeesSimilar(new BN(fee), partialFee)) {
+					return {
+						partialFee: fee.toString(),
+					};
+				}
+			}
+		}
+
+		// Check the Event::Deposit for the treasury pallet
+		const treasuryEvent = this.findEvent(events, 'treasury', Event.deposit);
+		if (treasuryEvent.length > 0 && treasuryEvent[0].data) {
+			const dataArr = treasuryEvent[0].data.toJSON();
+			if (Array.isArray(dataArr)) {
+				const fee = (dataArr as Array<number>)[0];
+
+				// The difference between values is 00.00001% or less so they are alike.
+				if (this.areFeesSimilar(new BN(fee), partialFee)) {
+					return {
+						partialFee: fee.toString(),
+					};
+				}
+			}
+		}
+
+		// Check Event::Deposit events for the balances pallet.
+		const depositEvents = this.findEvent(events, 'balances', Event.deposit);
+		if (depositEvents.length > 0) {
+			let sumOfFees = new BN(0);
+			depositEvents.forEach(
+				({ data }) =>
+					(sumOfFees = sumOfFees.add(new BN(data[data.length - 1].toString())))
+			);
+
+			// The difference between values is 00.00001% or less so they are alike.
+			if (this.areFeesSimilar(sumOfFees, partialFee)) {
+				return {
+					partialFee: sumOfFees.toString(),
+				};
+			}
+		}
+
+		return {
+			partialFee: partialFee.toString(),
+			error: 'Could not find a reliable fee within the events data.',
```

Comment:
I know I suggested returning an error at some point, but I am wondering whether, especially called from the `/blocks` endpoint, it makes sense or not? I'm not sure how often we'll find the fee in events vs having to use the queryInfo result?

We could choose to acknowledge that we'll use events if possible, but if no events are available, it's not an error. We should tell the user whether we managed to use an event or not, so I wonder about a return type here that is just `string | undefined` (either we found the fee in events or not), and then in `getPartialFeeInfo` a return type like:

```
{
    partialFee: string,
    partialFeeFromEvent?: boolean
    error,
    dispatchClass
}
```

Just a thought tho!